[Team15 - BE] 1주차 PR

Seohyoung opened the pull request:
안녕하세요 team15 우디 트리입니다!

### 클래스 명 네이밍 문제
- 현재 구현한 로직에서는 필요한 데이터만 추출해서 response 할 수 있도록 DTO를 만들어서 구현하였습니다.
- 저희가 명명, 사용한 DTO가 사실 DAO인지 아니면 Entity인지 궁금합니다.

### [ API 관련 참고사항 ]
- 제공된 Mock API를 기준으로 만들었습니다
- 현재 밥찬들(main, soup, side)에 대한 url 매핑을 아래와 같이 구현했습니다.
  - `@GetMapping("/main")`, `@GetMapping("/soup")`, `@GetMapping("/side")`
  -  원래는 `@PathVariable`로 밥찬들을 포괄할 수 있게 메소드 한 개만 구현을 해놨으나, `@GetMapping("/best")` 와 겹치는 문제가 발생하여 각각의 메소드로 나누어준 상태입니다.
  - 프론트와 협의해서 이후 ItemController에 공통 url(ex, `/dish/main`)을 만들어서 `@GetMapping("dish/{section_name}")`로 통합할 예정입니다. 이 점 참고해주세요!!

jihye-woo committed:
[#6] build: Initialize Spring Boot project

- 스프링 부트 프로젝트 기본 셋업
- 프로젝트 상위 폴더에 .gitignore 파일 추가

jihye-woo committed:
[#6] build: Add datasource properties

- application.properties에 datasource 설정 추가

choitree committed:
[#15] Feat : Create schema.sql , data.sql

- 테이블 컬럼 구조에 맞춰 스키마 및 데이터 생성

choitree committed:
[#16] Feat : Create Domain

- 도메인 생성

choitree committed:
[#18]Feat : Create DishRepository

- DishRepository 생성 및 테스트 코드 추가

choitree committed:
[#19]Feat : Create DishService

- DishService 생성 및 테스트 코드 추가

choitree committed:
[#18]Feat : Create DishService

- DishService 생성 및 테스트 코드 추가

choitree committed:
[#18]Feat : Create DishController

- DishController 생성 및 테스트 코드 추가

choitree committed:
[#20]Feat : Create DishController

- DishController 생성 및 테스트 코드 추가

choitree committed:
Merge branch 'BE/feature/16/domain' of https://github.com/jihye-woo/sidedish into BE/feature/16/domain

choitree committed:
[#17]Feat : Create DTO

- DishDTO, BestResponseDTO 생성

jihye-woo committed:
[#17] feature: Set default value at delivery table

- delivery 테이블 필드 delivery_type, delivery_fee, delivery_condition의 디폴트 값 설정
- data.sql에 수정된 스키마에 맞추어 쿼리문 수정

jihye-woo committed:
[#17] refactor: Renaming DishDTO to ItemDTO

- Mockup API 네이밍을 참고해서, DishDTO에서 ItemDTO로 클래스명 수정

jihye-woo committed:
[#17] refactor: Delete unused class

- 사용하지 않는 클래스 SectionResponseDTO.java 삭제

jihye-woo committed:
[#17] feature: Add custom query at DishRepository

- DishRepository::findById, DishRepository::findAll 메소드를 오버라이드
  - 커스텀 쿼리를 사용하여, 클라이언트의 요청에 맞는 데이터(ItemDTO)를 추출

jihye-woo committed:
[#17] feature: Add showDishsBySection method

- 특정 section에 속한 dish 데이터를 가져오는 기능 구현
  - DishController::showDishsBySection 메소드 구현
  - DishService::showDishsBySection 메소드 구현
  - DishRepository::findAllBySection 구현 및 커스텀 쿼리 추가

jihye-woo committed:
[#17] refactor: Renaming Dish to Item

- Mockup API의 네이밍에 맞게 Dish라는 명칭을 Item으로 수정

jihye-woo committed:
[#17] refactor: Remove unused method

- 불필요한 메소드 ItemController::showDishByHash 삭제

jihye-woo committed:
[#18] Create BestRepository

- BestRepository 생성

jihye-woo committed:
[#20] feature: Create BestController

- BestController 생성

jihye-woo committed:
[#20] feature: Split ItemController::showDishsBySection

- 특정 세션에 대한 메소드들로 분리

jihye-woo committed:
[#19] feature: Create BestService

- BestService 생성

jihye-woo committed:
[#16] feature: Create Best

- Best 테이블과 매핑되는 Best 엔티티 클래스 생성

jihye-woo committed:
[#16] build: Include gradle-wrapper.jar file

- 배포 시 jar을 빌드하기 위해, gradle-wrapper.jar 파일 origin으로 커밋

jihye-woo committed:
[#16] feature: Delete Event table

- DB에서 이벤트 테이블을 삭제
- DB 테이블 변경 사항을 클래스와 커스텀 쿼리에 적용

choitree committed:
Chore :Change (application.properties) datasource.url

-배포 url로 properties 변경

jihye-woo committed:
[#16] feature: Add DishBest table in DB

- dish_best 테이블을 추가하면서 data.sql과 schema.sql에도 변경사항을 반영

jihye-woo committed:
[#16] feature: Add BestDTO class

- 클라이언트가 요청한 json 데이터 형식에 맞춰 BestDTO 구현

jihye-woo committed:
[#16] feature: Add BestDish class

- 추가한 DishBest 테이블과 매핑되는 클래스 구현

jihye-woo committed:
[#16] feature: Add features for best API

- best API와 관련된 기능 구현
  - /best -> 모든 best 목록 가져오기
  - /best/{category_id} -> 특정 best 정보 가져오기

jihye-woo committed:
Merge remote-tracking branch 'origin/BE/feature/16/domain' into BE/feature/16/domain

jihye-woo committed:
Merge pull request #30 from jihye-woo/BE/feature/16/domain

[BE] 1차 PR

ksundong commented on .gitignore:
```diff
@@ -0,0 +1,176 @@
+Temporary Items
+.apdisk
+
+### react ###
```
전체 공통으로 쓰이는 건가요?
react에 관련된 ignore가 들어있네요?

choitree replied:
리액트 관련된 내용 삭제하였습니다~

ksundong commented on backend/.gitignore:
```diff
@@ -0,0 +1,214 @@
+Temporary Items
+.apdisk
+
+### react ###
```
여기도 React에 관련된 것이 들어있네요..?

ksundong commented on backend/src/main/java/com/team15/sidedish/controller/BestController.java:
```diff
@@ -0,0 +1,29 @@
+        this.bestService = bestService;
+    }
+
+    @GetMapping("/best")
```
중복제거가 이뤄지지 않았네요.
또, PR에서도 설명하셨듯, best로 시작하는 것은 적절치 않겠죠~

choitree replied:
/best로 두고, main,soup,side를 /section/main , /section/soup, section/side로 하는 방식으로 변경하였습니다.

ksundong commented on backend/src/main/java/com/team15/sidedish/controller/BestController.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @GetMapping("/best")
+    public List<BestDTO> showAllBestDishes() {
```
DTO의 사용 좋네요.

ksundong replied:
그런데, 이렇게 리턴하면 다음과 같은 형태가 될겁니다.

``` json
[
  {},
  {}
]
```

이 경우에는 API의 요구사항이 변경될 경우 Breaking Change가 발생하게 됩니다.

```json
"bestDishes": [
  {}
]
```

이런 형태는 추가적인 요구사항이 발생해도, Client에 영향을 미치지 않게되겠죠?

choitree replied:
Dion이 말씀해주신 내용이 약간 이해되지 않아서 제가 이해한 바로 다시 여쭤보겠습니다.
저희가 리턴하는 타입은 List 형식으로 리턴이 되는데, 그런 경우 
Breaking Change (API가 사라지거나 return type 혹은 parameters가 변경되는 경우)에 
유지보수 측면으로 좋은 구조가 아니다라고 이해했는데, 제가 이해한 바가 맞나요??

우선 해당 리턴 타입은 HashMap으로 변경했습니다.

ksundong replied:
API 가 사라지는건 Breaking Change지만, 이런 경우는 없다고 보셔도 됩니다.
Parameter가 변경되는 경우(예를 들어, bestDishes 말고도 다른 정보가 필요하다던가 하는 경우)에 유지보수 하기가 좋아집니다.
그리고 Map으로 리턴하지말고 DTO를 생성해서 반환해주세요~

choitree replied:
DTO 타입으로 변경하였습니다.

ksundong commented on backend/src/main/java/com/team15/sidedish/controller/BestController.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @GetMapping("/best/{categoryId}")
+    public BestDTO showSingleBestDish(@PathVariable Integer categoryId) {
```
Integer로 받을 필요가 있을까요?

choitree replied:
Long으로 변경하였습니다!

ksundong commented on backend/src/main/java/com/team15/sidedish/controller/ItemController.java:
```diff
@@ -0,0 +1,44 @@
+
+    @GetMapping("/dish/{hash}")
+    public ItemDTO showSingleDish(@PathVariable String hash) {
+        logger.info("Show single dish by hash");
```
debug가 적절해보이네요.

ksundong commented on backend/src/main/java/com/team15/sidedish/controller/ItemController.java:
```diff
@@ -0,0 +1,44 @@
+
+    }
+
+    @GetMapping("/main")
+    public List<ItemDTO> showMainDishes(){
+        return itemService.showDishsBySection("main");
+    }
+
+    @GetMapping("/soup")
+    public List<ItemDTO> showSoupDishes(){
+        return itemService.showDishsBySection("soup");
+    }
+
+    @GetMapping("/side")
+    public List<ItemDTO> showSideDishes(){
+        return itemService.showDishsBySection("side");
+    }
```
이 구조는 서로가 아무런 관계가 없이 url로 연결되어있습니다.
또한, 하드코딩도 되어있습니다. 개선해볼까요?

choitree replied:
/section/{sectionName}으로 매핑을 변경하고 리팩토링하였습니다!

ksundong commented on backend/src/main/java/com/team15/sidedish/domain/Best.java:
```diff
@@ -0,0 +1,28 @@
+
+public class Best {
+    @Id
+    private Integer categoryId;
```
Integer를 사용해주신 이유가 있나요?

ksundong replied:
클래스 설계가 조금 독특해보입니다.

choitree replied:
Long 타입으로 변경했습니다

ksundong commented on backend/src/main/java/com/team15/sidedish/domain/Best.java:
```diff
@@ -0,0 +1,28 @@
+    private Integer categoryId;
+    private String name;
+
+    public Best() { }
```
```suggestion
    public Best() {}
```

ksundong replied:
빈 생성자를 정의해주신 이유는 무엇인가요?

choitree replied:
getter 와 함께 작성했었는데, 우디가 빈 생성자는 JPA와 연관되었다고 알려주어서 학습 후 삭제하였습니다~~

ksundong replied:
저희가 쓰는 것은 Spring Data JDBC인데, 한 번 확인해주세요!

ksundong replied:
아하, 삭제하신거군요.

ksundong commented on backend/src/main/java/com/team15/sidedish/domain/Delivery.java:
```diff
@@ -0,0 +1,57 @@
+
+    private String dishHash;
+
+    public Delivery(){ }
```
```suggestion
    public Delivery() {}
```
리포맷 해주세요~

ksundong commented on backend/src/main/java/com/team15/sidedish/domain/Delivery.java:
```diff
@@ -0,0 +1,57 @@
+        return dishHash;
+    }
+
+    //    private List<DeliveryType> parsingDeliveryType(String delivery) {
+//        return Arrays.stream(delivery.split(“,”))
+//                .map(DeliveryType::valueOf)
+//                .collect(Collectors.toList());
+//    }
```
사용하지 않는 코드는 바로 제거해주세요.

choitree replied:
네네!!🙂

ksundong commented on backend/src/main/java/com/team15/sidedish/domain/Delivery.java:
```diff
@@ -0,0 +1,57 @@
+    private Long id;
+
+    private boolean isMonday;
+    private String deliveryType;
```
뭔가 Type하니까 Enum일 것 같은데, 아닌가보네요~

choitree replied:
넵 현재는 Enum이 아닌데, 리팩토링 할 여유가 된다면 변경해보겠습니다🙂

ksundong commented on backend/src/main/java/com/team15/sidedish/domain/Delivery.java:
```diff
@@ -0,0 +1,57 @@
+
+    private boolean isMonday;
+    private String deliveryType;
+    private String deliveryFee;
```
Fee가 String 타입인 이유가 있나요?

choitree replied:
처음에는 2500원 으로 명시하려고 String으로 사용했는데, int타입으로 값을 전달해주는 것으로 변경하고
DB에만 내용이 반영되었네요~ 모두 Integer 타입으로 변경했습니다!

ksundong commented on backend/src/main/java/com/team15/sidedish/domain/Dish.java:
```diff
@@ -0,0 +1,74 @@
+    private String sectionName;
+    private int stock;
+
+    public Dish(){ }
```
```suggestion
    public Dish() {}
```

ksundong commented on backend/src/main/java/com/team15/sidedish/domain/Dish.java:
```diff
@@ -0,0 +1,74 @@
+    private String topImage;
+    private String title;
+    private String description;
+    private int nPrice;
+    private int sPrice;
```
네이밍 이렇게 써주시는거 확실한가요?

choitree replied:
제공되는 API에서 변형하지 않고, 최대한 구조를 유지하려고 하다 보니 문제가 계속 발생한 것 같습니다ㅠㅠ
위에서 리뷰주신 내용은 조금 더 고민해보고 변경하도록 하겠고,
네이밍도 API 기준으로 작성했는데, 프론트와 협의 후 변경하겠습니다🥲

ksundong replied:
네네~

ksundong commented on backend/src/main/java/com/team15/sidedish/domain/Image.java:
```diff
@@ -0,0 +1,21 @@
+        this.imageUrl = imageUrl;
+        this.dishHash = dishHash;
+    }
+
+
+
+
```
nit: 빈 라인도 정리해주는 습관을 들여주세요.

choitree replied:
넵!

ksundong commented on backend/src/main/java/com/team15/sidedish/domain/ItemRepository.java:
```diff
@@ -0,0 +1,23 @@
+
+public interface ItemRepository extends CrudRepository<ItemDTO, String> {
+
+    @Query("SELECT di.`hash` AS `detail_hash`, di.`top_image` AS `image`, de.`delivery_type`, di.`title`, di.`description`, di.`n_price`, di.`s_price`, di.`badge` from dish di JOIN delivery de on di.`hash` = de.`dish_hash` where de.`dish_hash`=:hash")
```
그냥 JOIN을 쓰면 어떤 JOIN이 되나요?
모두 다 작성해주신건 잘하셨어요.

choitree replied:
Inner Join이 됩니다!~ delivery  테이블과 dish 테이블은 1:1 관계라서 Join하였습니다!

ksundong replied:
그러면 JOIN 보다는 INNER JOIN을 써주는게 좋을 것 같습니다~

choitree replied:
INNER JOIN으로 변경하였습니다.

ksundong commented on backend/src/main/java/com/team15/sidedish/domain/ItemRepository.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    List<ItemDTO> findAll();
+
+    @Query("select di.`hash` AS `detail_hash`, di.`top_image` AS `image`, de.`delivery_type`, di.`title`, di.`description`, di.`n_price`, di.`s_price`, di.`badge` from dish di JOIN delivery de on di.`hash` = de.`dish_hash` where di.`section_name` LIKE CONCAT('%', :section, '%')")
```
section은 사용자의 입력을 받아서 결정되는건가요?
아니라면 과연 LIKE를 써줘야했을지 의문이 듭니다.

choitree replied:
제공된 데이터들을 살펴보니 section이 main이면서 cource인 경우도 있었습니다. 하지만 그런 경우는 1~2개의 케이스밖에 없어서
section을 테이블로 분리하기는 애매하여 String으로 해당하는 section을 모두 받았고
section에 해당하는 정보가 전달되도록 LIKE를 사용했습니다.

ksundong replied:
아하 알겠습니다~

ksundong commented on backend/src/main/java/com/team15/sidedish/domain/ItemRepository.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.List;
+import java.util.Optional;
+
+public interface ItemRepository extends CrudRepository<ItemDTO, String> {
```
왜 DTO가 Repository로 왔을까요?

choitree replied:
DAO로 변경하였습니다!

ksundong commented on backend/src/main/java/com/team15/sidedish/dto/BestDTO.java:
```diff
@@ -0,0 +1,44 @@
+        this.items = items;
+    }
+
+    public static BestDTO of(Best best, List<ItemDTO> items){
+        return new BestDTO(best.getCategoryId(), best.getName(), items);
+    }
```
정적 팩토리 메서드의 사용 좋네요.

ksundong commented on backend/src/main/java/com/team15/sidedish/dto/BestDish.java:
```diff
@@ -0,0 +1,19 @@
+package com.team15.sidedish.dto;
+
+public class BestDish {
+    private String dish_hash;
+    private Integer category_id;
```
네이밍은 자바 네이밍 컨벤션을 지켜주세요.

choitree replied:
수정하였습니다

ksundong commented on backend/src/main/java/com/team15/sidedish/dto/ItemDTO.java:
```diff
@@ -0,0 +1,87 @@
+    private Integer nPrice;
+    @JsonProperty("s_price")
+    private Integer sPrice;
+    private String[] badge;
```
배열을 써주신 이유가 있나요?

choitree replied:
badge가 2개인 경우는 2가지 케이스밖에 없기 때문에, event를 별도로 분리하기 보다는 sectionName처럼 한 줄에 두가지 경우를 모두 가지고 있도록 사용하였습니다. DB에서 ,로 작성하고 배열 타입으로 선언하니 , 기준으로 2개가 분리되기에 데이터 가공이 쉽도록 배열로 사용하였습니다.

ksundong commented on backend/src/main/java/com/team15/sidedish/dto/ItemDTO.java:
```diff
@@ -0,0 +1,87 @@
+    @JsonProperty("detail_hash")
+    private String detailHash;
+    private String image;
+    private String alt;
```
alt는 무슨의미인가요?

choitree replied:
저희가 주어진 API를 그대로 유지하려고 하다보니, 불필요한 부분이더라도 동일한 구조로 만들기 위해서 
작성한 부분과 변수명들이 있습니다. 
금요일 데모 이후에 API를 변경해서 사용한 조들을 보면서 저희도 필요한 부분과 네이밍에 대해서 수정하는 작업이 필요했다고 깨닫고 후회하는 중입니다ㅠㅠ

alt 또한 이 과정의 산물(?)인데요, 결국 title과 동일한 내용이 들어가기 때문에 생성자에서 alt에 title을 대입해주는 방식으로 사용하고 있지만, alt를 굳이 전달하지 않더라도 프론트에서 사용할 수 있는 부분이라서 해당 변수는 삭제하도록 하겠습니다

ksundong commented on backend/src/main/java/com/team15/sidedish/dto/ItemDTO.java:
```diff
@@ -0,0 +1,87 @@
+    private String title;
+    private String description;
+
+    @JsonProperty("n_price")
+    private Integer nPrice;
+    @JsonProperty("s_price")
+    private Integer sPrice;
```
n_price는 뭐고 s_price는 뭐에요?
이 이름 정말 확실한가요?

choitree replied:
n_price -> normalPrice
s_price -> specialPrice 로 변경했습니다.
@JsonProperty는 카멜케이스가 아닌 스네이크 케이스로 json 키값이 전달되도록 변경하기 위해 사용했습니다!

ksundong commented on backend/src/main/java/com/team15/sidedish/service/BestService.java:
```diff
@@ -0,0 +1,41 @@
+    public List<BestDTO> showAllBestDishes() {
+        return bestRepository.findAll()
+                .stream()
+                .map(best -> showSingleBestDish(best.getCategoryId()))
```
매번 DB를 호출하는 형태인가요?
이건 N + 1 쿼리 아닐까요?

choitree replied:
N+1 쿼리를 해결하기 위해서 고민해보았지만, 시간 부족과 해결 방법이 떠오르지 않아 아직 해결하지 못하였습니다ㅠㅠㅠㅠ
우선 구현이 되지 않았던 부분을 추가하고 해당 부분은 추후에 리팩토링해보도록 하겠습니다!

ksundong commented on backend/src/main/java/com/team15/sidedish/service/BestService.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public BestDTO showSingleBestDish(Integer categoryId) {
+        List<String> bestDish = bestRepository.findByCategoryId(categoryId);
+        Best best = bestRepository.findById(categoryId).orElseThrow(IllegalArgumentException::new);
+
+        List<ItemDTO> items = bestDish
+                .stream()
+                .map(hash -> itemService.showSingleDish(hash))
+                .collect(Collectors.toList());
```
음.. 이 코드는 잘 이해가 되지 않네요.
일단 Best라는 클래스가 존재해야하는지 한 번 질문을 던져볼게요.
Best의 name에는 어떤 것이 들어가나요?

그리고 bestRepository에서 무엇을 찾는지 명확하지도 않네요. `findByCategoryId`면 `Best` 들이 나와야 할 것 같은데 `String`이 나오네요.

그리고 매 번 다른 서비스를 호출해서 굉장히 비효율적으로 쿼리가 발생할 것 같은데
이를 해결해보는게 숙제겠네요.

`IllegalArgumentException`도 그냥 던지지 마시고, Custom Exception의 도입을 고려해봐주세요.

choitree replied:
현재 저희 데이터 구조는 Best 테이블과 Dish 테이블이 별도로 구성되어 있고, Best와 1:1, Dish와 N:1 관계인 Best_Dishes 테이블이 존재합니다. 
Best 클래스는 Best 테이블과 동일하게 categoryId, name에 대한 정보만 존재하고
BestDTO는 Best_Dishes의 데이터를 응답받을 수 있는 구조입니다.

우선 Best_Dishes의 해당 카테고리 아이디가 일치하는 hash값을 List로 받아오고
Best의 이름과 id를 받아온 후
hash값에 맞는 dish를 List 형태로 담아
정적 메소드를 통해 BestDTO의 클래스 형태에 맞게 데이터를 리턴하는 방식으로 구현하였습니다.

디온에게 설명하면서 리팩토링이 가능할 것 같다는 사실을 알게되었네요..!!
이 숙제는..! 추후에 꼭꼭 리팩토링을 해보도록 하겠습니다

ksundong commented on backend/src/main/java/com/team15/sidedish/service/ItemService.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public ItemDTO showSingleDish(String hash) {
+        return itemRepository.findById(hash).orElseThrow(IllegalArgumentException::new);
```
여기도 Custom Exception 생각해봐주세요~

ksundong commented on backend/src/main/resources/application.properties:
```diff
@@ -0,0 +1,7 @@
+spring.datasource.url=jdbc:mysql://15.165.15.210:3306/sidedish
+spring.datasource.username=sidedish
+spring.datasource.password=sidedish
```
로컬환경도 아닌데 이렇게 노출되는 것은 매우 위험합니다.

ksundong reviewed (changes requested):
안녕하세요! Tree & Woody! 리뷰어 Dion 입니다~
두 번째 프로젝트가 생각보다 어렵게 느껴지는 것 같습니다. 설계도 그렇고, 이미 내려진 API의 형태를 그대로 따라가는 형태가 나오고 있네요.

자, 질문드리겠습니다. 우리가 만약에 위치 정보를 얻기 위해서 외부의 API를 사용한다면, 그 API가 위도 경도를 latitude, hardness 같은 식으로 준다고 하더라도, 그대로 사용하실건가요?
또, API는 결국 Interface입니다. 변경사항에 의해 Client가 변경되는 일은 최대한 줄여야 합니다. 변경 가능성이 적은 설계를 항상 염두에 두는 것이 중요합니다.

## 질문에 대한 답변

>현재 구현한 로직에서는 필요한 데이터만 추출해서 response 할 수 있도록 DTO를 만들어서 구현하였습니다.  
>저희가 명명, 사용한 DTO가 사실 DAO인지 아니면 Entity인지 궁금합니다.

자, 일단 용어를 사용하기에 앞서서 용어를 이해하고 사용해야합니다. 자신이 사용한 용어조차 모른다면, 대화가 불가능하지 않을까요?

- DTO는 무엇인가요?
- DAO는 무엇인가요?
- Entity는 무엇인가요?
- Domain은 무엇일까요?

이 대답을 제가 해드리는건 도움이 되지 않을 것 같습니다. 직접 찾아보시고, 코멘트 남겨주세요.
그 후에 무엇이 무엇인지에 대해서 더 얘기를 해볼 수 있을 것 같습니다.

## 리뷰

- API Path가 아쉽습니다. 더 좋은 Path를 고민해봅시다.
- Breaking Change가 발생하지 않을 수 있는 API 설계도 고민해봅시다.
- `Best`라는 클래스의 존재 의의가 무엇인지 궁금합니다. 단순 name을 저장하기 위한건가요?
- naming에 있어서 아쉬운 부분이 있습니다. 꼭 개선해주셨으면 좋겠네요.
- 작성하신 코드의 N + 1 쿼리문제는 개선할 방법을 꼭 찾아주세요.
- Custom Exception을 정의해보고, 예외 핸들링까지 해봅시다.

## 참고사항 리뷰

> 원래는 @PathVariable로 밥찬들을 포괄할 수 있게 메소드 한 개만 구현을 해놨으나, @GetMapping("/best") 와 겹치는 문제가 발생하여 각각의 메소드로 나누어준 상태입니다.

왜 겹칠까요? 겹치는 이유를 반드시 생각해봅시다. 겹쳐선 안되는 것들 아닌가요? 잘 생각해봅시다. 놓친게 있을겁니다.

> 프론트와 협의해서 이후 ItemController에 공통 url(ex, /dish/main)을 만들어서 @GetMapping("dish/{section_name}")로 통합할 예정입니다. 이 점 참고해주세요!!

Item이라는 이름이 뭔가 의아하게 느껴지네요. ItemController에서 Dishes를 주는건가요?

---

프로젝트 진행하시느라 고생 많으십니다. 리뷰 한 번 읽어보시고, 적용하고 싶은 부분 적용해주세요~
궁금한 점, 도움이 필요한 부분, 이해가 되지 않는 부분 등은 코멘트 꼭 남겨주세요!

jihye-woo committed:
[#32] chore : .gitignore 파일 수정

- react와 관련된 gitignore 설정 제거

jihye-woo committed:
[#32] refactor : BestController 공통 url 묶기

- BestController의 클래스 레벨에 공통 url 설정
  - @RequestMapping("/best")

jihye-woo committed:
[#32] refactor : Wrapping response to Map<K, V>

- 클라이언트에 보낼 응답 객체를 Map<>에 래핑하여 전달

choitree committed:
chore : Add bestdishes data to data.sql

- data.sql에 Best API에 있는 데이터 추가

jihye-woo committed:
[#32] refactor : Change n_price as nullable

- schema.sql에서 n_price를 Null로 변경
- schema.sql에서 s_price를 Not Null로 변경

jihye-woo committed:
[#32] refactor : Remove unused log message

- ItemController에서 사용하지 않는 log 메시지 제거

jihye-woo committed:
[#32] refactor : Change Id type from Integer to Long

- id 타입을 Integer에서 Long으로 변경
- 디폴트 생성자 제거
- 주석처리 된 메소드 제거

jihye-woo committed:
[#32] refactor : Change price type to Integer

- Delivery 클래스의 deliveryFee, deliveryCondition는 price에 대한 내용이므로 Integer 타입으로 변경

jihye-woo committed:
[#32] refactor : Rename nPrice and sPrice

- Dish 클래스의 nPrice와 sPrice 변수명을 조금 더 명확한 이름으로 수정
  - nPrice ➡️ normalPrice
  - sPrice ➡️ specialPrice
- 불필요한 기본 생성자 제거
- int 기본형으로 선언된 변수 타입을 Integer 레퍼 클래스 타입으로 수정

jihye-woo committed:
[#32] refactor : Remove unused class BestDish

- 사용하지 않는 클래스 BestDish 제거

jihye-woo committed:
[#32] refactor : Rename nPrice and sPrice

- Dish 클래스의 nPrice와 sPrice 변수명을 조금 더 명확한 이름으로 수정
 - nPrice ➡️ normalPrice
 - sPrice ➡️ specialPrice

- 바뀐 변수명에 맞춰 커스텀 쿼리문에 alias 컬럼 추가

jihye-woo committed:
[#32] feature : Add event table

- event 테이블을 db에 추가하고 schema.sql과 data.sql에 변경사항 반영

jihye-woo committed:
[#32] refactor : Rename Item to Dish

- Item 대신 Dish로 네이밍 변경

choitree committed:
[#32]Chore: update query, join to outer join

- Event가 없는 경우도 있어, outer join으로 변경

choitree committed:
[#32]Chore : update query join to outer join

- Event는 데이터가 없는 경우가 있어서, outer join으로 변경

choitree committed:
[#32]Refactor : delete duplicate method

-중복되는 메소드 제거 및 리팩토링

choitree committed:
[#32]Refactor : delete duplicate method

-중복되는 메소드 제거 및 리팩토링

choitree committed:
[#32] Feature : Create BestDishesDTO to wrapping Best list response

choitree committed:
#32] Refactor : Change response wrapping class from Map to Custom DTO

-Map으로 리턴하는 메소드를 BestDishDTO로 리턴하도록 변경

choitree committed:
#32] Refactor : Change response wrapping class from Map to Custom DTO

-Map으로 리턴하는 메소드를 BestDishDTO로 리턴하도록 변경

choitree committed:
[#32] Chore : Change table name dish_best to best_dishes, and column names

- 스키마 인덴팅 추가 및 외래키 명 변경
- 스키마 변경에 따른 insert문 테이블명, 컬럼명 변경

choitree committed:
[#32] Chore : Change ItemController class name to DishController

-Item 클래스 이름 변경에 따라서 ItemController 클래스를 DishController로 변경

jihye-woo committed:
[#32] refactor : Add @JsonProperty annotation

- 보내는 json 필드명을 snake case로 통일

jihye-woo committed:
[#32] feature : Create SectionDishesDTO for wrapping response

- SectionDishesDTO 클래스 구현
  - 각 밥찬(section)에 대한 응답 데이터를 SectionDishesDTO에 래핑

- DishController::showDishesBySectionName에 적용
  - SectionDishesDTO로 래핑된 응답 데이터를 클라이언트에게 리턴

jihye-woo committed:
[#32] refactor : Rename items to dishes at BestDTO

- BestDTO items 변수명 변경
  - DB 테이블 명과의 통일성을 위해 items 대신 dishes 사용

jihye-woo committed:
[#32] refactor : Add static factory method at BestDishesDTO

- BestDishesDTO::of 구현
  - BestDishesDTO 내부에 정적 팩토리 메소드 생성
  - BestController::showAllBestDishes 메소드에 적용

- BestDishesDTO의 생성자를 private으로 변경
  - 외부에서 DTO의 생성자에 직접 접근하는 대신, 정적 팩토리 메소드를 통해서 객체를 생성하도록 함

choitree committed:
Chore : Add all data in data.sql

-data.sql에 존재하는 데이터 모두 추가

(best는 이미지가 열리지 않아서,
 best에 연관된 image테이블 db는 존재하지 않고,
 top_image도 안열리는 상태)

jihye-woo committed:
chore : Add insert query in data.sql

- data.sql에 주어진 데이터들을 DB에 인서트하는 쿼리문을 추가

choitree committed:
Chore : fix all data in data.sql

-data.sql 오타 수정

choitree committed:
Chore : change event rows to one row

-event가 2개씩 있는 로우를 한개로 병합

choitree committed:
Chore : change query table name and column names

- best_dishes 테이블명과 best_category_id 컬럼명 변경

choitree committed:
Chore : Delete unused Class

choitree committed:
[#38]Feature : Make Detail Feature

-Detail 정보를 보는 로직 구현
 -제공된 API에 맞추어 Detail 정보에 hash와 data property를 출력하고
 -data property에 전달되는 값을 출력하도록 구현

choitree committed:
[#38]Feature : Make Detail Feature

-Detail 정보를 보는 로직 구현
 -제공된 API에 맞추어 Detail 정보에 hash와 data property를 출력하고
 -data property에 전달되는 값을 출력하도록 구현

jihye-woo committed:
build : Resolve deploy error

jihye-woo committed:
build : Remove automate sql initialize mode

jihye-woo committed:
build : Change query string to lowercase

ksundong commented:
@choitree @jihye-woo 
변경중이신건가요? commit은 계속 올라오는데 Review요청이 없어서 여쭤봅니다.

choitree committed:
style : add space at comment

-주석에 띄어쓰기 추가

choitree committed:
[#41] Change DataDTO data type

- FE 요청대로 디테일 데이터 필드 변경
 - price 배열을 각각 전달, normalPrice null일 경우 필드에서 제외
 - title 정보 추가
 - badge 정보 추가

choitree committed:
[#41] Change DataDTO data type

- FE 요청대로 디테일 데이터 필드 변경
 - price 배열을 각각 전달, normalPrice null일 경우 필드에서 제외
 - title 정보 추가
 - badge 정보 추가

choitree committed:
[#41] Change DishDTO class name to DishDAO

- DishDTO 클래스 이름을 DishDAO로 변경

choitree committed:
[#41] Change DishDTO class name to DishDAO

- DishDTO 클래스 이름을 DishDAO로 변경

choitree commented:
디온 재리뷰 요청이 지연된 점 죄송합니다.ㅠㅠ
아직 구현되지 못한 부분과 배포에 시간이 할애되고 있어서 디온이 주신 리뷰 내용에 온전히 집중하기가 어려웠습니다.
리뷰주신 내용에 코멘트를 작성하기 위해 다시 코드를 뜯어(?)보면서 조금 더 간결해질 수 있다는 사실을 인지하였지만 
우선 작동이 되도록 구현하는 저희의 목표를 우선으로 진행하다보니 많은 부분을 놓친 것 같습니다
금일 배포 작업을 진행하고, 명일에 가능하다면 배포와 무관한 브랜치를 만들어 리팩토링하도록 하겠습니다.

디온이 신경써주셨는데 코멘트가 많이 지연된 점 양해 부탁드립니다 :sob:

choitree committed:
[#41] Add JsonInclude annotation at normalPrice, badge

- normalPrice, badge에 NON_NULL 어노테이션 추가

choitree committed:
Chore : add UNSIGNED which is Dish table, stock column in schema.sql

- Dish 테이블에 stock컬럼에 음수가 들어갈 수 없도록 UNSIGNED 설정 추가

choitree committed:
Feature : Create custom Exception classes

- 사용자 정의 Exception 추가

choitree committed:
Feature : Add stock update method in Dish class

- 주문 시 재고 변동이 반영되도록, update 메소드 구현

choitree committed:
Feature : Add stock update logic

- 주문 시 재고가 변동될 수 있도록 로직 구현
